Sources/PackageCollectionsSigning/PackageCollectionSigning.swift

```diff
@@ -0,0 +1,195 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Foundation
+
+import PackageCollectionsModel
+
+public struct PackageCollectionSigning {
+    public typealias Model = PackageCollectionModel.V1
+
+    private static let minimumRSAKeySizeInBits = 2048
+
+    let certPolicy: CertificatePolicy
+
+    public init() {
+        self.init(certPolicy: NoopCertificatePolicy())
+    }
+
+    init(certPolicy: CertificatePolicy) {
+        self.certPolicy = certPolicy
+    }
+
+    /// Signs package collection using the given certificate and key.
+    ///
+    /// - Parameters:
+    ///   - collection: The package collection to be signed
+    ///   - certChainPaths: Paths to all DER-encoded certificates in the chain. The certificate used for signing
+    ///                     must be the first in the array.
+    ///   - certPrivateKeyPath: Path to the private key (*.pem) of the certificate
+    ///   - jsonEncoder: The `JSONEncoder` to use
+    ///   - callback: The callback to invoke when the signed collection is available.
+    public func sign(collection: Model.Collection,
+                     certChainPaths: [URL],
+                     certPrivateKeyPath: URL,
+                     jsonEncoder: JSONEncoder = JSONEncoder(),
+                     callback: @escaping (Result<Model.SignedCollection, Error>) -> Void) {
+        guard !certChainPaths.isEmpty else {
+            return callback(.failure(PackageCollectionSigningError.emptyCertChain))
+        }
+
+        do {
+            // Check that the cert is valid before we do anything
+            let certChainData = try certChainPaths.map { try Data(contentsOf: $0) }
+            let certChain = try certChainData.map { try Certificate(derEncoded: $0) }
+            self.certPolicy.validate(certChain: certChain) { result in
+                switch result {
+                case .failure(let error):
+                    return callback(.failure(error))
+                case .success(let isCertChainValid):
```

Comment:
It's probably better to throw error instead of returning false. https://github.com/apple/swift-package-manager/pull/3242/commits/a0e7f10ee6a6898e1828a3fd1ff5b6e41fdbb910